specs/interop/predeploys.md

```diff
@@ -0,0 +1,260 @@
+      tstore(CHAINID_SLOT, _id.chainid)
+    }
+
+    bool success = SafeCall.call({
+      _target: _target,
+      _value: msg.value,
+      _calldata: _msg
+    });
+
+    require(success);
+}
+```
+
+### `Identifier` Getters
+
+The `Identifier` MUST be exposed via `public` getters so that contracts can call back to authenticate
+properties about the `_msg`.
+
+## L2ToL2CrossDomainMessenger
+
+| Constant          | Value                                        |
+|-------------------|----------------------------------------------|
+| Address           | `0x4200000000000000000000000000000000000024` |
+| `MESSAGE_VERSION` | `uint256(0)`                                 |
+| `INITIAL_BALANCE` | `type(uint248).max`                          |
+
+The `L2ToL2CrossDomainMessenger` is a higher level abstraction on top of the `CrossL2Inbox` that
+provides features necessary for secure transfers of `ether` and ERC20 tokens between L2 chains.
+Messages sent through the `L2ToL2CrossDomainMessenger` on the source chain receive both replay protection
+as well as domain binding, ie the executing transaction can only be valid on a single chain.
+
+### `relayMessage` Invariants
+
+- Only callable by the `CrossL2Inbox`
+- The `Identifier.origin` MUST be `address(L2ToL2CrossDomainMessenger)`
+- The `_destination` chainid MUST be equal to the local chainid
+- The `CrossL2Inbox` cannot call itself
+
+### Message Versioning
+
+Versioning is handled in the most significant bits of the nonce, similarly to how it is handled by
+the `CrossDomainMessenger`.
+
+```solidity
+function messageNonce() public view returns (uint256) {
+    return Encoding.encodeVersionedNonce(nonce, MESSAGE_VERSION);
+}
+```
+
+### Transferring Ether in a Cross Chain Message
+
+The `L2ToL2CrossDomainMessenger` MUST be initially set in state with an ether balance of `INITIAL_BALANCE`.
+This initial balance exists to provide liquidity for cross chain transfers. It is large enough to always have
+ether present to dispurse while still being able to accept inbound transfers of ether without overflowing.
+The `L2ToL2CrossDomainMessenger` MUST only ensure all invariants are held before transferring out any ether.
+
+The `L2CrossDomainMessenger` is not updated to include the `L2ToL2CrossDomainMessenger` functionality because
+there is no need to introduce complexity between the L1 to L2 messages and the L2 to L2 liquidity.
+
+### Interfaces
+
+The `L2ToL2CrossDomainMessenger` uses a similar interface to the `L2CrossDomainMessenger` but
+the `_minGasLimit` is removed to prevent complexity around EVM gas introspection and the `_destination`
+chain is included instead.
+
+#### Sending Messages
+
+The initiating message is represented by the following event:
+
+```solidity
+event SentMessage(bytes message) anonymous;
+```
+
+The `bytes` are an ABI encoded call to `relayMessage`. The event is defined as `anonymous` so that no topics
+are prefixed to the abi encoded call.
```

Comment:
The reason we want no topics is to save gas or to avoid having to slice off the first 32 bytes when parsing? With only a single event name emitted by this contract an anonymous one seems ok, but we probably want an event in `relayMessage` and then it becomes harder to distinguish the two (e.g. now you rely on number of topics to distinguish them which doesn't sound ideal)

Reply:
The identifier points to a log where the log must be encoded like so:

```
msg := make([]byte, 0)
for _, topic := range log.Topics {
    msg = append(msg, topic.Bytes()...)
}
msg = append(msg, log.Data...)
```

This is because sometimes there is important information in the topics, but then you need an adapter contract to map a topic to a 4byte selector. It is because this `msg` is the calldata passed into a contract

Reply:
> The reason we want no topics is to save gas or to avoid having to slice off the first 32 bytes when parsing? With only a single event name emitted by this contract an anonymous one seems ok, but we probably want an event in `relayMessage` and then it becomes harder to distinguish the two (e.g. now you rely on number of topics to distinguish them which doesn't sound ideal)

The `Identifier` points exactly to a single log and if the identifier doesn't exactly match the serialized log and the tx is included in a block, its considered an invalid block

Reply:
> The reason we want no topics is to save gas or to avoid having to slice off the first 32 bytes when parsing? With only a single event name emitted by this contract an anonymous one seems ok, but we probably want an event in `relayMessage` and then it becomes harder to distinguish the two (e.g. now you rely on number of topics to distinguish them which doesn't sound ideal)

The `Identifier` points exactly to a single log and if the identifier doesn't exactly match the serialized log and the tx is included in a block, its considered an invalid block